text/0000-sizeof-alignof-offsetof-typeof.md

```diff
@@ -0,0 +1,93 @@
+
+These yield a value of type `usize` and may be used in any expression contexts.
+
+A more formal grammar:
+```
+expr          : ... | sizeof_expr | alignof_expr | offsetof_expr
+sizeof_expr   : "sizeof" "(" type ")"
+alignof_expr  : "alignof" "(" type ")"
+offsetof_expr : "offsetof" "(" struct_field "in" struct_type ")"
+struct_field  : ident | integer
+```
+
+Examples:
+```rust
+struct MyStruct {
+    pub field1: i32,
+    pub field2: f64
+}
+
+sizeof(MyStruct) == 16;
+alignof(MyStruct) == 8;
+offsetof(field2 in MyStruct) == 8;
+```
+
+# Alternatives
+
+- Beef up rustc's constant folder, so it can see though \<raw-pointer-deref\>-\<field-select\>-\<get-address-of\> combos.  It would then be possible to express `sizeof` and `offsetof` in terms of raw pointer operations.  For example: `macro_rules! offsetof(($f:ident in $t:ty) => (unsafe{ &(*(0 as *const $t)).$f as usize }));`
+Cons: The implementations of these operators via pointer arithmetic use dubious tricks like dereferencing a null pointer, and may become broken by future improvements in LLVM optimizations.  Also, this approach seems to be falling out of fashion even in C/C++ compilers (I guess for the same reason): `alignof`is now a standard operator in C++11; the `offsetof` hasn't been standartized yet, but both GCC and clang implement it as a custom built-in.
+
+- Implement a limited form of compile-time function evaluation (CTFE) by hard-coding knowledge of intrinsics such as `size_of<T>()` and `align_of<T>()` into the constant folder.
+Cons:
+  - There is no syntax in Rust for referring to a struct field, other than in assignment context, so the equivalent of C++'s `sizeof Struct::field` cannot be expressed in it.
+  - For `offsetof` this is an even bigger problem, because its core functionality requires referring to the field, whose offset we want to determine.  We could represent it as a string, i.e. `offset_of<T>(field: &str)`, but then what do we do in non-const contexts?  If we allow arbitrary string parameters, we'd have to bring back runtime reflection.  If we don't, it would be the only function in the language that *requires* a literal parameter.
+
+# Unresolved questions
+
+- Should we allow usage of private struct fields with `offsetof` or only of the public ones?
+- Can these operators be made to work in array length expressions (and eventually in numeric type parameters,- when these get implemented)?  The trouble is that array lengths need to be known in the type checking phase, however the physical data type layouts are known only to LLVM in trans phase.
+
+
+## Syntax bikeshedding
+
+The above syntax looks very similar to regular function calls.  Some alternatives:
+- Macro-like syntax: `sizeof!(MyStruct); alignof!(f64); offsetof!(field2 in MyStruct))`.
+Pros: distinct from regular functions; "non-standard" syntax is par for the course in macros; it's immediately obvious that these are evaluated at compile time.
```

Comment:
We actually have some compiler-builtin macroses already, so adding a few new ones would not be too bad IMO. And they can still be expanded to their literal value.

That being said, I think I’d prefer the macro version.